pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Wai.hs

```diff
@@ -199,36 +209,99 @@ sendResponse cb waitAct = do
 
 liveReq :: Ship -> TVar LiveReqs -> RAcquire e (Word64, STM RespAct)
 liveReq who vLiv = mkRAcquire ins del
- where
-  ins = atomically (newLiveReq who vLiv)
-  del = atomically . rmLiveReq vLiv . fst
-
-app
-  :: HasLogFunc e
-  => e
-  -> Ship
-  -> TVar LiveReqs
-  -> (Word64 -> ReqInfo -> STM ())
-  -> (Word64 -> STM ())
-  -> KingSubsite
-  -> W.Application
-app env who liv inform cancel sub req respond =
+  where
+    ins = atomically (newLiveReq who vLiv)
+    del = atomically . rmLiveReq vLiv . fst
+
+app ::
+  HasLogFunc e =>
+  e ->
+  Ship ->
+  TVar LiveReqs ->
+  (Word64 -> ReqInfo -> STM ()) ->
+  (Word64 -> STM ()) ->
+  EyreSite ->
+  KingSubsite ->
+  W.Application
+app env who liv inform cancel site sub req respond = do
+  met <- maybe (error "bad method") pure (cookMeth req)
+  let getp = met == H.GET
+      scryFlow =
+        onException
+          (runEyreSite site req respond)
+          normalFlow
+      normalFlow =
+        runEyreSite
+          (normalSite env who liv inform cancel)
+          req
+          respond
   case W.pathInfo req of
-    ("~_~":_) -> runKingSubsite sub req respond
-    _ ->
-      runRIO env $ rwith (liveReq who liv) $ \(reqId, respApi) -> do
-        bod <- io (toStrict <$> W.strictRequestBody req)
-        met <- maybe (error "bad method") pure (cookMeth req)
-
-        let adr = reqAddr req
-            hdr = W.requestHeaders req
-            url = reqUrl req
-
-        atomically $ inform reqId $ ReqInfo adr met url hdr bod
-
-        try (sendResponse respond respApi) >>= \case
-          Right rr  -> pure rr
-          Left  exn -> do
-            atomically (cancel reqId)
-            logError $ display ("Exception during request" <> tshow exn)
-            throwIO (exn :: SomeException)
+    ("~_~" : _) -> runKingSubsite sub req respond
+    _ -> if getp then scryFlow else normalFlow
+
+normalSite ::
+  HasLogFunc e =>
+  e ->
+  Ship ->
+  TVar LiveReqs ->
+  (Word64 -> ReqInfo -> STM ()) ->
+  (Word64 -> STM ()) ->
+  EyreSite
+normalSite env who liv inform cancel = ES $ \req respond ->
```

Comment:
As proposed above, move these into the where block in `app` so you don't have to take in all these arguments. Also get rid of the EyreSite type. Since we no longer need to pass these whole "sites" though the stack, we don't need a newtype to document them.